jlm/util/Statistics.hpp

```diff
@@ -64,8 +75,164 @@ class Statistics
+   * @return the measurement's value
+   */
+  [[nodiscard]] size_t
+  GetUInteger(const std::string & name) const;
+
+  /**
+   * Gets the floating point value the measurement with the given \p name.
+   * Requires the measurement to exist and be of type double.
+   * @return the measurement's value.
+   */
+  [[nodiscard]] double
+  GetDouble(const std::string & name) const;
+
+  /**
+   * Retrieves the full list of measurements
+   */
+  [[nodiscard]] const MeasurementList &
+  GetMeasurements() const;
+
+  /**
+   * Checks if a timer with the given \p name exists.
+   * @return true if the timer exists, false otherwise.
+   */
+  [[nodiscard]] bool
+  HasTimer(const std::string & name) const noexcept;
+
+  /**
+   * Gets the timer with the given \p name, it must exist.
+   * @return the timer.
+   */
+  [[nodiscard]] const util::timer &
+  GetTimer(const std::string & name) const;
+
+  /**
+   * Retrieves the full list of timers
+   */
+  [[nodiscard]] const TimerList &
+  GetTimers() const;
+
+protected:
+  /**
+   * Adds a measurement, identified by \p name, with the given value.
+   * Requires that the measurement doesn't already exist.
+   * Measurements are listed in insertion order.
+   */
+  void
+  AddString(std::string name, std::string value);
+
+  void
+  AddInteger(std::string name, int64_t value);
+
+  void
+  AddUInteger(std::string name, size_t value);
+
+  void
+  AddDouble(std::string name, double value);
+
+  // Mutable version of @see GetMeasurement
+  [[nodiscard]] Measurement &
+  GetMeasurement(const std::string & name);
+
+  /**
+   * Removes the measurement with the given \p name, if it exists.
+   * @return true if the measurement existed and was removed, false otherwise.
+   */
+  bool
+  RemoveMeasurement(const std::string & name);
+
+  /**
+   * Creates a new timer with the given \p name and starts it.
+   * Requires that the timer does not already exist.
+   * @return a reference to the timer
+   */
+  util::timer &
+  StartTimer(std::string name);
+
+  /**
+   * Stops and resets the timer with the given \p name.
+   * Requires that the timer exists.
+   * @return a reference to the timer
+   */
+  util::timer &
+  ResetTimer(const std::string & name);
+
+  /**
+   * Stops the timer with the given \p name.
+   * Requires the timer to exist, and be running.
+   * @return a reference to the timer
+   */
+  util::timer &
+  StopTimer(const std::string & name);
+
+  /**
+   * Resumes the timer with the given \p name.
+   * Requires the timer to exist, and be stopped.
+   * @return a reference to the timer
+   */
+  util::timer &
+  ResumeTimer(const std::string & name);
+
+  // Mutable version of @see GetTimer
+  [[nodiscard]] util::timer &
+  GetTimer(const std::string & name);
+
+  /**
+   * Removes the timer with the given \p name, if it exists.
+   * @return true if the timer existed and was removed, false otherwise
+   */
+  bool
+  RemoveTimer(const std::string & name);
```

Comment:
This interface feels a little bit inconsequent to me:

1. I would remove `RemoveTimer()` from it. Do we have a use case for this method?
2. You are offering several methods reflecting the operations on a timer, but at the same time also return the underlying timer for each of those methods, which could be used to perform the same operations.

Alternative interface suggestion:

```
util::timer &
AddTimer(std::string name);

[[nodiscard]] util::timer &
GetTimer(const std::string & name);
```

And then rely on the timer interface for performing the necessary operations. In addition, I would delete the copy constructor and copy assignment operator, as it would be way too easy to do this mistake:

`auto timer  = GetTimer("myTimer")` instead of `auto & timer = GetTimer("myTimer")`.